### PR TITLE
added option for mixed sg-full grid

### PR DESCRIPTION
**Pull request by mkstoyanov**
Please review the [developer documentation](https://github.com/project-asgard/asgard/wiki/developing)
on the wiki of this project that contains help and requirements.

## Proposed changes

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

## What type(s) of changes does this code introduce?
Added option to mix sparse grid and full grid by breaking the dimensions into two groups and using sparse grid within each group and dense between the two groups.

- [ ] Bugfix
- [x] New feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

### Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Checklist

_Put an x in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- [x] this PR is up to date with current the current state of 'develop'
- [x] code added or changed in the PR has been clang-formatted
- [ ] this PR adds tests to cover any new code, or to catch a bug that is being fixed
- [x] documentation has been added (if appropriate)

### CODE REVIEW

File: src/program_options.cpp [hunk 1]
```diff
@@ -39,6 +39,11 @@ parser::parser(int argc, char const *const *argv)
       clara::detail::Opt(max_level,
                          "integer >= all starting levels")["-m"]["--max_level"](
           "Maximum hierarchical levels (resolution) for adaptivity") |
+      clara::detail::Opt(
+          mixed_grid_group,
+          "integer < num dimensions")["-mgg"]["--mixed_grid_group"](
+          "Dimension group size for first spars grid in the mixed sparse-dense "
```
Comment (quantumsteve) [inline src/program_options.cpp]:
```suggestion
          "Dimension group size for first sparse grid in the mixed sparse-dense "
```

File: src/adapt_tests.cpp [hunk 2]
```diff
@@ -119,21 +119,22 @@ TEMPLATE_TEST_CASE("adapt - 1d, scattered coarsen/refine", "[adapt]", double,
...
-  auto const do_adapt_levels = parser::DEFAULT_DO_ADAPT;
-  auto const adapt_threshold = adapt_thresh;
+  auto const degree           = 3;
+  auto const level            = 4;
+  auto const pde_choice       = PDE_opts::continuity_1;
+  auto const num_dims         = 1;
+  auto const cfl              = parser::DEFAULT_CFL;
+  auto const use_full_grid    = parser::DEFAULT_USE_FG;
+  auto const max_level        = parser::DEFAULT_MAX_LEVEL;
+  auto const mixed_grid_group = -1;
```
Comment (quantumsteve) [inline src/adapt_tests.cpp]:
```suggestion
  auto const mixed_grid_group = parser::DEFAULT_MIXED_GRID_GROUP;
```

File: src/adapt_tests.cpp [hunk 3]
```diff
@@ -144,21 +145,22 @@ TEMPLATE_TEST_CASE("adapt - 2d, all zero", "[adapt]", double, float)
...
-  auto const do_adapt_levels = parser::DEFAULT_DO_ADAPT;
-  auto const adapt_threshold = adapt_thresh;
+  auto const degree           = 2;
+  auto const level            = 5;
+  auto const pde_choice       = PDE_opts::continuity_2;
+  auto const num_dims         = 2;
+  auto const cfl              = parser::DEFAULT_CFL;
+  auto const use_full_grid    = parser::DEFAULT_USE_FG;
+  auto const max_level        = parser::DEFAULT_MAX_LEVEL;
+  auto const mixed_grid_group = -1;
```
Comment (quantumsteve) [inline src/adapt_tests.cpp]:
```suggestion
  auto const mixed_grid_group = parser::DEFAULT_MIXED_GRID_GROUP;
```

File: src/adapt_tests.cpp [hunk 4]
```diff
@@ -174,21 +176,22 @@ TEMPLATE_TEST_CASE("adapt - 3d, scattered, contiguous refine/adapt", "[adapt]",
...
-  auto const do_adapt_levels = parser::DEFAULT_DO_ADAPT;
-  auto const adapt_threshold = adapt_thresh;
+  auto const degree           = 4;
+  auto const level            = 4;
+  auto const pde_choice       = PDE_opts::continuity_3;
+  auto const num_dims         = 3;
+  auto const cfl              = parser::DEFAULT_CFL;
+  auto const use_full_grid    = parser::DEFAULT_USE_FG;
+  auto const max_level        = parser::DEFAULT_MAX_LEVEL;
+  auto const mixed_grid_group = -1;
```
Comment (quantumsteve) [inline src/adapt_tests.cpp]:
```suggestion
  auto const mixed_grid_group = parser::DEFAULT_MIXED_GRID_GROUP;
```

File: src/adapt_tests.cpp [hunk 5]
```diff
@@ -229,21 +232,22 @@ void test_initial(parser const &problem, std::string const &gold_filepath)
...
-  auto const do_adapt_levels = true;
-  auto const adapt_threshold = adapt_thresh;
+  auto const degree           = 4;
+  auto const level            = 3;
+  auto const pde_choice       = PDE_opts::diffusion_1;
+  auto const num_dims         = 1;
+  auto const cfl              = parser::DEFAULT_CFL;
+  auto const use_full_grid    = parser::DEFAULT_USE_FG;
+  auto const max_level        = parser::DEFAULT_MAX_LEVEL;
+  auto const mixed_grid_group = -1;
```
Comment (quantumsteve) [inline src/adapt_tests.cpp]:
```suggestion
  auto const mixed_grid_group = parser::DEFAULT_MIXED_GRID_GROUP;
```

File: src/adapt_tests.cpp [hunk 6]
```diff
@@ -259,21 +263,22 @@ TEMPLATE_TEST_CASE("initial - diffusion 2d", "[adapt]", double, float)
...
-  auto const do_adapt_levels = true;
-  auto const adapt_threshold = adapt_thresh;
+  auto const degree           = 3;
+  auto const level            = 2;
+  auto const pde_choice       = PDE_opts::diffusion_2;
+  auto const num_dims         = 2;
+  auto const cfl              = parser::DEFAULT_CFL;
+  auto const use_full_grid    = parser::DEFAULT_USE_FG;
+  auto const max_level        = parser::DEFAULT_MAX_LEVEL;
+  auto const mixed_grid_group = -1;
```
Comment (quantumsteve) [inline src/adapt_tests.cpp]:
```suggestion
  auto const mixed_grid_group = parser::DEFAULT_MIXED_GRID_GROUP;
```

File: src/kronmult_tests.cpp [hunk 7]
```diff
@@ -20,7 +20,7 @@ parser get_parser(PDE_opts const pde_choice,
                   int const memory_limit)
 {
   return parser(pde_choice, starting_levels, degree, parser::DEFAULT_CFL,
-                parser::DEFAULT_USE_FG, parser::DEFAULT_MAX_LEVEL,
+                parser::DEFAULT_USE_FG, parser::DEFAULT_MAX_LEVEL, -1,
```
Comment (quantumsteve) [inline src/kronmult_tests.cpp]:
```suggestion
                parser::DEFAULT_USE_FG, parser::DEFAULT_MAX_LEVEL, parser::DEFAULT_MIXED_GRID_GROUP,
```

File: src/moment_tests.cpp [hunk 8]
```diff
@@ -31,7 +31,7 @@ TEMPLATE_TEST_CASE("Multiwavelet", "[transformations]", double, float)
   auto const adapt_threshold      = 0.5e-1;
 
   parser const parse(pde_choice, levels, degree, cfl, full_grid,
-                     parser::DEFAULT_MAX_LEVEL, num_steps, use_implicit,
+                     parser::DEFAULT_MAX_LEVEL, -1, num_steps, use_implicit,
```
Comment (quantumsteve) [inline src/moment_tests.cpp]:
```suggestion
                     parser::DEFAULT_MAX_LEVEL, parser::DEFAULT_MIXED_GRID_GROUP, num_steps, use_implicit,
```

File: src/moment_tests.cpp [hunk 9]
```diff
@@ -68,7 +68,7 @@ TEMPLATE_TEST_CASE("CreateMomentReducedMatrix", "[moments]", double, float)
   auto constexpr tol_factor = get_tolerance<TestType>(100);
 
   parser const parse(pde_choice, levels, degree, cfl, full_grid,
-                     parser::DEFAULT_MAX_LEVEL, num_steps, use_implicit,
+                     parser::DEFAULT_MAX_LEVEL, -1, num_steps, use_implicit,
```
Comment (quantumsteve) [inline src/moment_tests.cpp]:
```suggestion
                     parser::DEFAULT_MAX_LEVEL, parser::DEFAULT_MIXED_GRID_GROUP, num_steps, use_implicit,
```

File: src/program_options.cpp [hunk 10]
```diff
@@ -39,6 +39,11 @@ parser::parser(int argc, char const *const *argv)
       clara::detail::Opt(max_level,
                          "integer >= all starting levels")["-m"]["--max_level"](
           "Maximum hierarchical levels (resolution) for adaptivity") |
+      clara::detail::Opt(
+          mixed_grid_group,
+          "integer < num dimensions")["-mgg"]["--mixed_grid_group"](
```
Comment (quantumsteve) [inline src/program_options.cpp]: The short option may only be one character
https://github.com/project-asgard/asgard/pull/510#discussion_r1126690429

File: src/program_options.hpp [hunk 11]
```diff
@@ -217,6 +219,7 @@ class parser
       int const degree_in = NO_USER_VALUE, double const cfl_in = DEFAULT_CFL,
       bool const use_full_grid_in          = DEFAULT_USE_FG,
       int const max_level_in               = DEFAULT_MAX_LEVEL,
+      int const mixed_grid_group_in        = -1,
```
Comment (quantumsteve) [inline src/program_options.hpp]:
```suggestion
      int const mixed_grid_group_in        = DEFAULT_MIXED_GRID_GROUP,
```
around line 185, add 
```
static auto constexpr DEFAULT_MIXED_GRID_GROUP = NO_USER_VALUE
```

File: src/program_options.hpp [hunk 12]
```diff
@@ -319,6 +324,8 @@ class parser
   int degree = NO_USER_VALUE;
   // max adaptivity level for any given dimension.
   int max_level = DEFAULT_MAX_LEVEL;
+  // number of dimensions to use for the tensor groups in mixed grid
+  int mixed_grid_group = -1;
```
Comment (quantumsteve) [inline src/program_options.hpp]:
```suggestion
  int mixed_grid_group = DEFAULT_MIXED_GRID_GROUP;
```

File: src/solver_tests.cpp [hunk 13]
```diff
@@ -22,7 +22,7 @@ parser get_parser(PDE_opts const pde_choice,
                   int const memory_limit)
 {
   return parser(pde_choice, starting_levels, degree, parser::DEFAULT_CFL,
-                parser::DEFAULT_USE_FG, parser::DEFAULT_MAX_LEVEL,
+                parser::DEFAULT_USE_FG, parser::DEFAULT_MAX_LEVEL, -1,
```
Comment (quantumsteve) [inline src/solver_tests.cpp]:
```suggestion
                parser::DEFAULT_USE_FG, parser::DEFAULT_MAX_LEVEL, parser::DEFAULT_MIXED_GRID_GROUP,
```

File: src/time_advance_tests.cpp [hunk 14]
```diff
@@ -136,7 +136,7 @@ TEMPLATE_TEST_CASE("time advance - diffusion 2", "[time_advance]", double,
     auto const full_grid = false;
     parser const parse(
         pde_choice, fk::vector<int>(std::vector<int>(num_dims, level)), degree,
-        cfl, full_grid, parser::DEFAULT_MAX_LEVEL, num_steps);
+        cfl, full_grid, parser::DEFAULT_MAX_LEVEL, -1, num_steps);
```
Comment (quantumsteve) [inline src/time_advance_tests.cpp]:
```suggestion
        cfl, full_grid, parser::DEFAULT_MAX_LEVEL, parser::DEFAULT_MIXED_GRID_GROUP, num_steps);
```
repeat change for other instances of `-1` in this file.

**Review by quantumsteve — CHANGES_REQUESTED** [review]
Let's create a default value that uses `parser::NO_USER_VALUE` and use that in place of `-1`.

**Comment by mkstoyanov** [comment]
I rewrote the parser, now we can make changes entry by entry. I also redid a bunch of the use cases.

Still ugly, but should be at least an improvement.

File: src/program_options.hpp [hunk 15]
```diff
@@ -362,6 +374,42 @@ class parser
   bool valid = true;
 };
 
+struct parser_mod
+{
+  enum parser_option_entry
```
Comment (quantumsteve) [inline src/program_options.hpp:379]: I think this can be an `enum class`

**Review by quantumsteve — APPROVED** [review]
Other than the comment about making an `enum` an `enum class`, LGTM.

File: src/program_options.hpp (hunk 15 above)
Comment (mkstoyanov) [reply src/program_options.hpp:379]: This is already inside a class and has to be addressed with `parser_mod::cli`, adding `enum class` will only add another layer of indirection: `parser_mod::parser_option_entry::cli` or moving the `enum` outside will the class will pollute the namespace with yet one more name.

Comment (quantumsteve) [reply src/program_options.hpp:379]: My goal was just to provide a bit more type-safety. Not worth adding adding at the expense of another level of indirection.

**Pull request merged**